### ISSUE TITLE
Memory leak in version 0.8.271

### ISSUE THREAD

**Issue by pjstadig** [cognitect/transit-clj#25]
My co-worker (Tanya Romankova) and I have discovered a memory leak in version 0.8.271 of transit-clj. We were attempting to encode several million items with transit (each with a separate writer), and eventually got a GC overhead limit OOME. After examining the heap dump it became clear that the issue was transit-java's WriteHandler cache.

Every time a writer is created a new hash map of default handlers is also created by calling cognitect.transit/default-write-handlers. Since the values of the hash map that is generated are reified objects, they have identity equality semantics. This means that no two hash maps that come out of cognitect.transit/default-write-handlers are equal:

```
(not= (cognitect.transit/default-write-handlers) (cognitect.transit/default-write-handlers))
=> true
```

The result is that the WriteHandler cache in transit-java accrues a new entry every time you create a writer.

We've created a repository that demonstrates the memory leak. The README contains details about our diagnosis as well as sample output.

https://github.com/pjstadig/transit-clj-0.8.271-memory-leak

**Comment by dchelimsky** [comment]
Hey @pjstadig thanks for this report. We've got a fix in place, just needs some minor tweaks and more testing and should have it out within the next couple of days.

**Comment by dchelimsky** [comment]
Just released transit-java-0.8.290 and transit-clj-0.8.275 with the fix for this. Here's the output https://github.com/pjstadig/transit-clj-0.8.271-memory-leak modified to use this new version:

```
$ lein with-profile transit-clj-0.8.275 run
(:repositories detected in user-level profiles! [:user]
See https://github.com/technomancy/leiningen/wiki/Repeatability)
(Warning: profile :dev-internal not found.)
(Warning: profile :datomic-pro not found.)
Max Heap: 50331648
transit-clj Version: 0.8.275
Encoded chunk of 1000 (238 ms)
Encoded chunk of 1000 (74 ms)
Encoded chunk of 1000 (37 ms)
Encoded chunk of 1000 (39 ms)
Encoded chunk of 1000 (38 ms)
Encoded chunk of 1000 (38 ms)
Encoded chunk of 1000 (37 ms)
Encoded chunk of 1000 (39 ms)
Encoded chunk of 1000 (35 ms)
Encoded chunk of 1000 (36 ms)
Encoded chunk of 1000 (33 ms)
Encoded chunk of 1000 (35 ms)
Encoded chunk of 1000 (27 ms)
Encoded chunk of 1000 (29 ms)
Encoded chunk of 1000 (34 ms)
Encoded chunk of 1000 (22 ms)
Encoded chunk of 1000 (25 ms)
Encoded chunk of 1000 (21 ms)
Encoded chunk of 1000 (29 ms)
Encoded chunk of 1000 (23 ms)
```

**Comment by dchelimsky** [comment]
And thank you so much, @pjstadig, for providing such a great bug report and a runnable repro. Model citizenship!
